### PR TITLE
Add support of Honeywell HumidIcon (I2C HIH series) Temperature & Humidity sensor

### PR DESCRIPTION
**Pull request by Benichou34**
# What does this implement/fix?
Add support of Honeywell HumidIcon (I2C HIH series) Temperature & Humidity sensor

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature:** [esphome/feature-requests/issues/1799]

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation:** esphome/esphome-docs#3357

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
```yaml
sensor:
  - platform: honeywell_hih_i2c
    update_interval: 60s
    temperature:
      name: "Temperature"
    humidity:
      name: "Humidity"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/honeywell_hih_i2c/honeywell_hih.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+#include "honeywell_hih.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace honeywell_hih_i2c {
+
+static const char *const TAG = "honeywell_hih";
+
+static const uint8_t RESQUEST_CMD[1] = {0x00};  // Measurement Request Format
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/honeywell_hih.cpp]:
```suggestion
static const uint8_t REQUEST_CMD[1] = {0x00};  // Measurement Request Format
```

File: esphome/components/honeywell_hih_i2c/honeywell_hih.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// https://prod-edam.honeywell.com/content/dam/honeywell-edam/sps/siot/en-us/products/sensors/humidity-with-temperature-sensors/common/documents/sps-siot-i2c-comms-humidicon-tn-009061-2-en-ciid-142171.pdf
+//
+
+#include "honeywell_hih.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace honeywell_hih_i2c {
+
+static const char *const TAG = "honeywell_hih";
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/honeywell_hih.cpp]:
```suggestion
static const char *const TAG = "honeywell_hih.i2c";
```

File: esphome/components/honeywell_hih_i2c/honeywell_hih.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/components/sensor/sensor.h"
+#include "esphome/components/i2c/i2c.h"
+
+namespace esphome {
+namespace honeywell_hih_i2c {
+
+class HONEYWELLHIComponent : public PollingComponent, public i2c::I2CDevice {
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/honeywell_hih.h]:
```suggestion
class HoneywellHIComponent : public PollingComponent, public i2c::I2CDevice {
```

File: esphome/components/honeywell_hih_i2c/honeywell_hih.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ public:
+  void dump_config() override;
+  float get_setup_priority() const override;
+  void loop() override;
+  void update() override;
+
+  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { temperature_sensor_ = temperature_sensor; }
+  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { humidity_sensor_ = humidity_sensor; }
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/honeywell_hih.h]:
```suggestion
  void set_temperature_sensor(sensor::Sensor *temperature_sensor) { this->temperature_sensor_ = temperature_sensor; }
  void set_humidity_sensor(sensor::Sensor *humidity_sensor) { this->humidity_sensor_ = humidity_sensor; }
```

File: esphome/components/honeywell_hih_i2c/honeywell_hih.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  // Check status bits
+  return ((data[0] & 0xC0) == 0x00);
+}
+
+void HONEYWELLHIComponent::measurement_timeout_() {
+  ESP_LOGE(TAG, "Honeywell HIH Timeout!");
+  this->measurement_running_ = false;
+  this->mark_failed();
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/honeywell_hih.cpp:64]: Should this mark as failed, or just set a warning/error status which can be cleared if it succeeds in a future update? Same in the other places that call `mark_failed`

File: esphome/components/honeywell_hih_i2c/sensor.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/sensor.py]:
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
```

File: esphome/components/honeywell_hih_i2c/sensor.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        cg.add(var.set_temperature_sensor(sens))
+
+    if CONF_HUMIDITY in config:
+        sens = await sensor.new_sensor(config[CONF_HUMIDITY])
```
Comment (jesserockz) [inline esphome/components/honeywell_hih_i2c/sensor.py]:
```suggestion
    if humidity_config := config.get(CONF_HUMIDITY):
        sens = await sensor.new_sensor(humidity_config)
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/honeywell_hih_i2c/honeywell_hih.cpp (hunk 5 above)
Comment (Benichou34) [reply esphome/components/honeywell_hih_i2c/honeywell_hih.cpp:64]: I think that if there is an I2c communication error or a timeout error, the device must be mark as failed.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**